### ISSUE TITLE
Enhancement: subtask

### ISSUE THREAD

**Issue by fvezzoli** [dmfs/opentasks#37]
please consider to add the support to the sub task.

**Comment by dmfs** [comment]
Apparently KOrganizer and Rainlendar support subtasks and there is planned support for ownCloud. Thanks Tom for reporting these.

**Comment by hazzl** [comment]
Mirakel also has subtasks. Although their CalDAV support seems a bit sketchy.

**Comment by untitaker** [comment]
See https://github.com/bitfireAT/davdroid/issues/34 about Mirakel with CalDAV.

**Comment by dmfs** [comment]
We're already working on this feature. Today I made a few more changes in the task provider to add some required fields (not sure how I could forget to add the UID field to the tasks table).

I also started to add support to CalDAV-Sync. The same update will also add sync of categories, attachments, comments and maybe attendees. But first we have to make sure that all users use the latest task version, otherwise CalDAV-Sync will crash, because relations depend on the UID field.

**Comment by untitaker** [comment]
What should be done with the current subtasks this app has, the GitHub-like ones in the DESCRIPTION?

**Comment by dmfs** [comment]
We leave it as it is, like "sub-tasks lite". For many users it's already sufficient to have this "checklist style" sub-task support. This is much more lightweight than real sub-tasks, implementation-wise and UI-wise. Also it still kind of works with servers/clients that don't support real sub-tasks (i.e most clients & servers). For distinction we call it "checklist" instead of sub-tasks.

Keeping the checklist is still nice to have. I'm using this feature for my grocery lists. Real sub-tasks would be like using a sledgehammer to crack a nut for this purpose

Sub-task are much more heavy weighted. They can have due dates, start dates, descriptions, attachments and all the other task properties on their own. In many cases that's completely unnecessary.

Of course we still can improve the checklist. Managing them from the details view would be nice. Changing the order too. But these are separate issues.

**Comment by Nebucatnetzer** [comment]
Will this make it work with the owncloud app tasks?
Or should I open a new issue for that?

**Comment by untitaker** [comment]
@Nebucatnetzer are you talking about general compatibility or this particular feature?

**Comment by Nebucatnetzer** [comment]
I'm not quite sure actually. I just noticed that when I have a list of tasks it doesn't convert to the owncloud format. In owncloud the list appears like this:

Stuff to buy
[ ] carrots
[ ] milk
[ ] coffee

**Comment by untitaker** [comment]
Yeah, ownCloud doesn't seem to support any kind of subtask: https://github.com/owncloud/tasks/issues/49

Currently you can edit the description field (e.g. replace the spaces between the brackets with an x), and it should sync just fine.

**Comment by Nebucatnetzer** [comment]
You can add list in owncloud. I don't know if that's new or in which way it is different however.
I tried it today for the first time.
https://www.2li.ch/owncloud/index.php/s/HOWWzPRGigCuVJP

**Comment by Nebucatnetzer** [comment]
Ah I think I found out how they do it.
It creates a new calendar.

**Comment by jkufner** [comment]
It appears this way in every other client and it is fine this way. These
are not actual subtasks as understood in advanced bug tracking systems.
These are simple checklists to complete given task. And their usecase is
different.

To implement fully-featured subtasks, there is another way: The ability
to link one task to another. The similar way as Redmine does. This way
user creates network of related tasks, where no task is actualy part of
another task, but he can specify such relation. Another tools can then
visualize such relations using simple dependency graph or Gantt graph.
Funny part on this approach is, that it does not make app more complex.
These relations are just simple bidirectional M:N relations. No need to
modify how app stores the tasks, just add one attribute and it is done.

**Comment by untitaker** [comment]
Linking tasks is exactly what this bug is about, see the linked RFC

**Comment by Nebucatnetzer** [comment]
Makes perfectly sense to do it the redmine way.
Does someone a Linux application which works the same way as the Tasks app?
Or can you point me into good direction to look for one?
Thunderbird doesn't seem to work with that format.

**Comment by dmfs** [comment]
@Nebucatnetzer I think @untitaker and @jkufner said it all. Our current checklist feature is just stores a simple ASCII checklist in the description. Even a task application that doesn't support subtasks can handle this. Real sub-tasks is a more powerful but also more complex thing.

I disagree with what @jkufner said:
> These relations are just simple bidirectional M:N relations. No need to modify how app stores the tasks, just add one attribute and it is done.

That's only partially true. There is a lot more than just storing an additional attribute, like
* adding indices to get a decent performance
* properly handle linked tasks that are not in the same task list
* maintaining consistency (e.g. when deleting a task that is a parent to other tasks)
* maintaining sanity (e.g. don't allow circular parent-child relationships)

Also there are different types of relationships:
* structural:
 * parent/child of
 * sibling of
* temporal
 * finish-start
 * start-finish
 * start-start
 * ...
* logical
 * blocks/blocked by
 * duplicates
 *  ...

@Nebucatnetzer see https://github.com/dmfs/tasks/issues/37#issuecomment-74844185 KOrganizer has been reported to support sub-tasks

**Comment by Nebucatnetzer** [comment]
Thanks for the help, I'll give a try :)

**Comment by jkufner** [comment]
Marten Gajda wrote, on 11.5.2015 17:21:
>   * adding indices to get a decent performance

Yes, but this depends on how many tasks we are talking about and what
queries are to be performed. If tasks are stored in Sqllite database,
this can be done by database engine with no additional costs in development.


>   * properly handle linked tasks that are not in the same task list

If URL of task is known, then we can easily link any task anywhere.
But I believe that allowing links within single calendar is good enough
and much simpler.

>   * maintaining consistency (e.g. when deleting a task that is a parent
>     to other tasks)

No need to bother with parent/child paradigm. That is the beauty of
simple relations. Just delete task and all its relations and we are done.

>   * maintaining sanity (e.g. don't allow circular parent-child
>     relationships)

Circular relations may not make sense, but it does not cause any
trouble. All algorithms must deal with possible circles because there
may be inconsistency. And user can easily fix such problem. Technically
circular dependencies are allowed, so no problem here.


> Also there are different types of relationships: [...]

There are no structual relations. Temporal and logical relations are
interpreted by visualising tool or something. But they do not have to be
interpreted at all. Specifying relations is enough in the first step.

The second step may be some nice algorithm which analyzes the graph and
tells you what can be done today (no blocked-by relations) and whether
or not is possible to make it until deadline. Also some progress bars
can be calculated and plots rendered using such data.

Point in these simple relations is, that app does not have to understand
them. But when it does, it can be awesome.

**Comment by untitaker** [comment]
>maintaining sanity (e.g. don't allow circular parent-child relationships)

I suppose you mean to look for a way to enforce this through the DB schema (instead of application code)?

---

@jkufner I suppose you're just brainstorming here, but anyways: I don't think any of these things are *conceptually* hard, but they definetly aren't *easy*, but a lot of (programming) work.

**Comment by jkufner** [comment]
Markus Unterwaditzer wrote, on 11.5.2015 20:55:
>     maintaining sanity (e.g. don't allow circular parent-child
>     relationships)
> 
> I suppose you mean to look for a way to enforce this through the DB
> schema (instead of application code)?

Not really. Simply ignoring this problem will not cause any immediate
trouble. There can (and should) be some verification when storing the
task and warning when it does not make sense, but it is not neccessary.


> @jkufner <https://github.com/jkufner> I suppose you're just
> brainstorming here, but anyways: I don't think any of these things are
> /conceptually/ hard, but they definetly aren't /easy/, but a lot of
> (programming) work.

Actually it is easy. I did it some time ago using MySQL database.

There are two ways to do this:

1. External index: Good option when SQL storage is used. All relations
are in dedicated table implementing good old M:N relation. Only
difference is, that relation is somewhat symetric, so joins need to
accept both directions (for example "blocked-by" and "blocks" are two
views on the same row in the SQL table). Something like `LEFT JOIN ref
ON ref.src = task.id OR ref.dst = task.id`.

2. Each task has list of related tasks and application must modify both
tasks at once. This approach is less dependent on storage and allows
referencing external tasks or tasks in different storages. I guess this
approach is better if cross-calendar references are to be implemented.
This way it is only about list of URLs with relation type in edited task
and updating referenced tasks when possible.

It is really that simple. That is what I like on this approach most.

**Comment by dmfs** [comment]
> I suppose you mean to look for a way to enforce this through the DB schema (instead of application code)?

No, that's probably impossible. But I think we should do certain sanity and/or consistency checks in the task content provider (i.e. the database management code).

@jkufner our goal is to handle 100s or even 1000s of tasks with reasonable performance. So we need to make sure we have proper indices. Also, it's not always guaranteed that we have access to a referenced task or a URL of it. In iCalendar related tasks are referred to by their UID not their URL.

Also, if you receive a task assignment via iMIP (i.e via Email), you may not have access to the related tasks. A related problem is the synchronization order. It's not guaranteed that a parent task was already synced when the children get synced. So you may have temporary inconsistencies that you have to resolve later on (when the parent task gets synced).

Btw. our database back-end already supports relations (see https://github.com/dmfs/task-provider/blob/master/src/org/dmfs/provider/tasks/TaskContract.java#L1346 , https://github.com/dmfs/task-provider/blob/master/src/org/dmfs/provider/tasks/taskhooks/RelationUpdaterHook.java and https://github.com/dmfs/task-provider/blob/master/src/org/dmfs/provider/tasks/handler/RelationHandler.java). We also have a CalDAV-Sync branch that can sync relations. It's "just" the UI part that's missing.

**Comment by Selaron** [comment]
New ownCloud app "Tasks Plus" declares to support subtask creation:
https://apps.owncloud.com/content/show.php/Tasks+Plus?content=170561

Would like to see something like that in your Android tasks app :-)

**Comment by 1337sup3rh4x0r** [comment]
Subtasks would be great! You planning anything?

**Comment by dmfs** [comment]
yes we do. The database (opentask-provider) already has some initial support for subtasks. Now it's mostly a matter of user-interface.

**Comment by IntelligyScience** [comment]
I know it is an old issue but i dont see that is resolved. 
I saw subtask are implemented. But tasks created in owncloud with subtask (using owncloud 9 subtasks are possible) will be shown as task and not as subtask of the main task. Is their any problem with owncloud?

Thank you

**Comment by b3niup** [comment]
+1 for @IntelligyScience question - is there any work planned to make OpenTasks compatible with OwnCloud Tasks or is it OwnCloud Tasks issue that subtask are handled differently?

**Comment by jowi24** [comment]
I also hit this issue, as subtask in the the tasks app for owncloud are not rendered correctly in this app. Sadly, this is a blocker for me.

**Comment by jfmcbrayer** [comment]
NextCloud seems to support subtasks, with the same issue as above with ownCloud -- subtasks appear in this app as top-level tasks (as is reasonable given lack of subtask support). This is not a blocker for me, as this app is the main way I'm going to interact with NextCloud tasks, and in the interim I can use checklists. But it would be Nice To Have.

**Comment by Braintelligence** [comment]
Yes, please consider implementing subtasks. The way opentasks behaves now, with using notes like "[ ] Subtask" is a bit strange if used together with NextCloud.

**Comment by welbert** [comment]
Any updates here??

**Comment by szotsaki** [comment]
May I ask if you have made any updates on adding substasks UI-related parts? Do you have any plans when will it be available?

Thank you!

**Comment by dmfs** [comment]
I've created #443 to track the subtasks UI implementation (step 1). We'll work on that next to the recurring tasks feature.

**Comment by szotsaki** [comment]
May I ask if you have any updates on adding substasks UI-related parts? Do you have any plans when will it be available?

Thank you!

**Comment by Grosskopf** [comment]
is there a wiki or something how i can get started with development in this repository? i would like to get this issue fixed and i'm not sure if i would like the approach of subtasks in the details window, why is it done in that way? :)

**Comment by christianlupus** [comment]
I am using the checklist style now quite some time. It is handy for small and dumb lists. But when it comes to more complex problems, a real subtask implementation like e.g. in nextcloud would be very nice.
For most of the time I would stick to the checklists, but when structuring larger projects, I need a more in-depth structure. I might e.g. need recursive sub tasks (aka subtasks of subtasks) which is no problem with the correct data structure underlying but with the checklists it is a pain in the ****. I need to unroll all the subtasks manually and it gets quite confusing if you a few dozen entries there.

So I would vote for a parallel solution: Keep the checklists for day-to-day usage and allow the real subtasks for heavy weighted work.

**Comment by szotsaki** [comment]
As a gentle ping may I ask how much time do you think it will take to add this feature to opentasks? Thank you!
